File: linked_lists/palindrome/sol3.py
import sys

sys.path.append("../")

from sll import sll

class result:
  def __init__(self, node, res):
    self._node = node
    self._res = res
    
  @property
  def node(self):
    return self._node
    
  @node.setter
  def node(self, node):
    self._node = node
  
  @property
  def res(self):
    return self._res
    
  @res.setter
  def res(self, res):
    self._res = res

def ispalrecur(nd, length):
  if nd is None or length <= 0:
    return result(nd, True)
  elif length == 1:
    return result(nd.next, True)
    
  res = ispalrecur(nd.next, length - 2)
  
  if not res.res or res.node is None:
    return res
  
  res.res = res.node.data == nd.data
  
  res.node = res.node.next
  
  return res

def ispal(slinked_list):
  return ispalrecur(slinked_list.head, slinked_list.size).res

    
if __name__ == "__main__":
  slinked_list1 = sll.sll()
  key = 0
  value = [7, 1, 8, 1, 7]
  for it in value:
    slinked_list1.addLast(key, it)
    key += 1
  
  print ispal(slinked_list1)
 

File: linked_lists/palindrome/sol3.cpp
#include "sll.h"

#include <iostream>
#include <vector>

class res {
public:
  res(node* nd, bool res) : d_nd(nd), d_res(res) {}
  
  node* getnd() { return d_nd; }
  bool getres() { return d_res; }
  
  void setnd(node* nd) { d_nd = nd; }
  void setres(bool res) { d_res = res; }
  
private:
  node* d_nd;
  bool d_res;
};

unsigned int linkedListSize(node* nd) {
  unsigned int res = 0;
  while ( nd != nullptr ) {
     ++res;
     nd = nd->next();
  }
  return res;
}

res* isPalindrome(node* nd, const int& size) {
  if ( nd == nullptr || size <= 0 ) {
     return new res(nd, true);
  } else if ( size == 1 ) {
     return new res(nd->next(), true);
  }
  
  res* Res = isPalindrome(nd->next(), size - 2);
  
  if ( !Res->getres() ) {
     return Res;
  }
  
  if ( Res->getnd()->data() == nd->data() ) {
     Res->setnd(Res->getnd()->next());
     return Res;
  }
  
  Res->setres(false);
  return Res;
}
  

int main() {
  sll slinked_list;
  std::vector<int> value{1, 2, 3, 2, 1, 1};
  int key = 0;
  for (auto& it : value) {
    if (key == 0) {
      slinked_list.insertHead(it, key);
    } else {
      slinked_list.insertLast(it, key);
    }
    ++key;
  }
  
  node* nd = slinked_list.head();
  while (nd) {
    std::cout << nd->data() << " ";
    nd = nd->next();
  }
  std::cout << std::endl;
   

  nd = slinked_list.head();
  int size = linkedListSize(nd);
  if ( isPalindrome(nd, size)->getres() ) {
     std::cout << "Palindrome";
  } else {
     std::cout << "is not";
  }
  
  std::cout << std::endl;
}
